index.d.ts

```diff
@@ -0,0 +1,252 @@
+      getCode(addressOrENSName: string, blockTag: string): Promise<string>
+      getStorageAt(addressOrENSName: string, position: string, blockTag: string): Promise<string>
+      sendTransaction(signedTransaction: string): Promise<string>
+      call(transaction: Transaction): Promise<string>
+      estimateGas(transaction: Transaction): Promise<utils.BigNumber>
+      getBlock(blockHashOrBlockTag: string): Promise<any>
+      getTransaction(transactionHash: string): Promise<Transaction>
+      getTransactionReceipt(transactionHash: string): Promise<any>
+      getLogs(filter: any): Promise<string[]>
+      getEtherPrice(): Promise<number>
+      resolveName(name: string): Promise<string>
+      lookupAddress(address: string): Promise<string>
+      on(eventName: string, listener: EventListener): void
+      once(eventName: string, listener: EventListener): void
+      emit(eventName: string): void
+      listenerCount(eventName: string): number
+      listeners(eventName: string): Promise<EventListener[]>
+      removeAllListeners(eventName: string): void
+      removeListener(eventName: string, listener: EventListener): void
+      resetEventsBlock(blockNumber: number): void
+      polling(value?: number)
+    }
+
+    class EtherscanProvider extends Provider {
+      public constructor(network: Network, apiKey: string)
+      perform(method: string, params: string[]): Promise<string>
+      getHistory(addressOrENSName: string, startBlock: number, endBlock: number): Promise<any[]>
+    }
+
+    class FallbackProvider extends Provider {
+      public constructor(providers: Provider[])
+      perform(method: string, params: string[]): Promise<string>
+    }
+
+    class JsonRpcProvider extends Provider {
+      public constructor(url: string, network?: Network | string)
+      send(method: string, params: string[]): Promise<string>
+      perform(method: string, params: string[]): Promise<string>
+    }
+
+    class InfuraProvider extends JsonRpcProvider {
+      public constructor(network: Network, apiAccessToken: string)
+    }
+
+    class Web3Provider extends JsonRpcProvider {
+      public constructor(web3provider: Web3CurrentProvider, network?: Network | string)
+      getSigner(address?: string): Web3Signer
+      listAccounts(): Promise<string[]>
+    }
+
+    class Web3Signer {
+      private constructor(provider: Web3Provider, address?: string)
+
+      getAddress(): Promise<string>
+      getBalance(blockTag: string): Promise<utils.BigNumber>
+      getTransactionCount(blockTag: string): Promise<number>
+      sendTransaction(transaction: Transaction): Promise<Transaction>
+      signMessage(message: string): Promise<string>
+      unlock(password: string): Promise<string>
+    }
+
+  }
+
+  export type Transaction = {
+    chainId: number
+    hash: string
+    from: string
+    to: string
+    data: any
+    nonce: utils.BigNumber
+    gasPrice: utils.BigNumber
+    gasLimit: utils.BigNumber
+    value: utils.BigNumber
```

Comment:
Optional.

Reply:
Consider having two different `Transaction` data structures.  One that describes a transaction from the blockchain where nothing is optional, and another that describes a transaction that is being submitted to a provider (which will fill in most things).  `wallet.sign` should take the non-optional one, while `wallet.sendTransaction` should take the optional one.